[RFC] Deferred device initialization - with callback

edersondisouza opened the pull request:
This is a shot at @fabiobaltieri idea at #67335 - using a callback to influence device initialization.

This basically works by registering a callback (this registration must be done before devices are initialized, using `SYS_INIT(EARLY)` for instance), one can check what is the next device to be initialised and decide if it should happen now or deferred. If deferred, devices need to be initialised manually via `device_init()`.

edersondisouza committed:
kernel: Device deferred initialization

Currently, all devices are initialized at boot time (following their
level and priority order). This patch introduces deferred
initialization: by registering a callback to be run before devices are
initialized (register it at SYS_INIT(EARLY), for instance), one can
tell Zephyr to not initialize a given device.

To initialize deferred devices, one has to call `device_init()`.

To reduce processing overhead for those who don't care about this
feature, it is shielded behind CONFIG_DEVICE_DEFER_INIT.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

edersondisouza committed:
tests/kernel/device: Add tests for device deferred init

Ensure that devices are not ready before calling `device_init()`, but
are after the call.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

fabiobaltieri commented on kernel/init.c:
```diff
@@ -280,36 +316,67 @@ static void z_sys_init_run_level(enum init_level level)
-				}
-			}
+#ifdef CONFIG_DEVICE_DEFER_INIT
+			if (device_before_init_registered_cb != NULL) {
```
How about using an iterable section of callbacks instead of a single one? The advantage is that the callbacks are registered at build time (no need to have SYS_INIT to register it) and you can have more than one of them. To implement that you'd have to define a struct with the callback as a member, then you can declare it in the linker with `ITERABLE_SECTION_ROM`, in the code with `STRUCT_SECTION_ITERABLE` and then iterate through them here with `STRUCT_SECTION_FOREACH`. I think it'd be fair to iterate through all callbacks for all devices, I'd expect the callback implementation to start with something like `if (dev != <statically defined device>) {return DEVICE_INIT;}` so there should not be much overhead.

fabiobaltieri commented on include/zephyr/linker/common-rom/common-rom-kernel-devices.ld:
```diff
@@ -22,6 +22,10 @@
 	ITERABLE_SECTION_ROM_NUMERIC(device, 4)
 
+#ifdef CONFIG_DEVICE_DEFER_INIT
+	ITERABLE_SECTION_ROM(z_device_before_init, 4)
```
cool! you should also add an entry in `cmake/linker_script/common/common-rom.cmake` for platforms that use `CONFIG_CMAKE_LINKER_GENERATOR`

fabiobaltieri commented:
Cool, thanks for putting this together, I think this is a good design and certainly covers the use cases I had in mind. Letting the others chime in for more opinions.

@henrikbrixandersen @nashif @gmarull @cfriedt

edersondisouza commented:
Hmmm... the qemu_x86_tiny happens because the callback function is "far", so a page fault happens when it is called, but mmu is not fully setup (it will be after POST_KERNEL, from a comment there). If I force it to be (via `__section__` and ld script trickery) near the init code, then it works... This is annoying.

michalfita commented:
Sorry for coming out of the blue.

Am I right this is unwritten **Option 5** for zephyrproject-rtos/zephyr#39896?

This option seem to not only be simpler but has benefit of taking control from the software logic flow perspective w/o amending the device tree to defer initialization. And fact of using the callback may be behind feature flag for different variants of the software. Plus, there a chance to make mistake of not declaring a device initialisation as deferred while the code actually calls for initialization expecting device to be deferred what puts burden on programmer to run through right checks. By saying so I don't ignore potential of Option 4 to control if device driver initialization is deferred without recompilation of the code.

fabiobaltieri commented:
> Am I right this is unwritten **Option 5** for #39896?

Yeah I suppose you could say so, not sure why this never came up in the original discussions, but that started long time ago and few things in the init support code changed in the meantime and I guess no one thought about the implications of that and focused on variations of the original proposals from 2021. Let's call it **Option A**. :-)

michalfita commented:
Great, can be **Option H** for me as long as it's merged & released sooner than later 😉

edersondisouza committed:
tests/kernel/device: Fix device deferred init on qemu_x86_tiny

Device initialization can happen before MMU is initialised. As the code
for the callback deferring mechanism may be far from initialisation
code, it may not be mapped yet. On qemu_x86_tiny this can cause an
exception like:

    ERROR   - SeaBIOS (version zephyr-v1.0.0-0-g31d4e0e-dirty-2020071[...]
    Booting from ROM..ASSERTION FAIL [page_frames_initialized] @ \
        WEST_TOPDIR/zephyr/kernel/mmu.c:1395
    	page fault at 0x401974f6 happened too early
    E: EAX: 0x00000001, EBX: 0x00009500, ECX: 0x00000000, EDX: 0x00000000
    E: ESI: 0x00000000, EDI: 0x00001000, EBP: 0x401935dc, ESP: 0x401935d8
    E: EFLAGS: 0x00000046 CS: 0x0008 CR3: 0x00171000
    E: call trace:
    E: EIP: 0x4013ef3b
    E:      corrupted? (bp=0x401935dc)
    E: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
    E: Current thread: 0x40193728 (unknown)
    E: Halting system

This patch introduces a __DEVICE_INIT attribute, which moves code
to just after the other device initialisers, thus avoiding this issue.
To avoid messing with linking on unaffected devices, this behaviour is
shielded by CONFIG_DEVICE_DEFER_CALLBACK_SECTION, currently only
selected for qemu_x86_tiny.

Signed-off-by: Ederson de Souza <ederson.desouza@intel.com>

andyross reviewed (dismissed):
Honestly I personally liked the simpler choice, but this looks great too.  Really any trick like this is going to be for some kludgey reason (because if there was dynamic state to honor during initialization at design time, it would be reflected in some way in the DTS tree to begin with), so elegance isn't ever going to be on the menu.

edersondisouza commented:
v2:
  - A fix for the qemu_x86_tiny issue (not related to the current qemu_x86_tiny issue. which has a fix at #68379)
  - Rebased.

fabiobaltieri commented on include/zephyr/device.h:
```diff
@@ -750,12 +750,60 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
+/**
+ * @brief Possible results for callback called before device initialization.
+ */
+enum device_before_init_result {
+	DEVICE_INIT,
+	DEVICE_DEFER_INIT
+};
```
Should we drop this and just return a boolean? maybe rename the field to something like `device_allow_init` or `device_block_init` and adjust accordingly. This is nice and unambiguous but these type of enum return are a bit uncommon in Zephyr and it somehow reminds me of those horrible `return HAL_SOMETHING_OK;` that few private SDK have. :-)

@gmarull or anyone else, any opinion? @cfriedt also suggested that in https://github.com/zephyrproject-rtos/zephyr/pull/67335#issuecomment-1887115782

cfriedt replied:
Hmm... yeah, this looks and smells like a `bool`

fabiobaltieri commented on include/zephyr/device.h:
```diff
@@ -750,12 +750,60 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
+ *  initialization
+ */
+#define DEVICE_BEFORE_INIT_CALLBACK(callback) \
+	STRUCT_SECTION_ITERABLE(z_device_before_init, _z_device_before_init##callback) = \
+				{.cb = callback}
```
nit: the codingstyle says brace in line for structure, so I think this wants to be

```c
        STRUCT_SECTION_ITERABLE(z_device_before_init, _z_device_before_init##callback) = { 
                .cb = callback,
        }
```

tbursztyka commented:
I personally don't like this approach, in comparison to #67335, it really is a hack. At least there is no DTS property. It anyway has the same limitation of as the other PR (added overhead, no dependency handling etc..).

fabiobaltieri commented:
> in comparison to #67335, it really is a hack

What do you find hacky about it?

> It anyway has the same limitation of as the other PR

The main thing that this approach has that the other one does not is the possibility of leaving a device to be initialized in the original spot in the sequence (which is validated against the dt hierarchy). Any approach that moves things around the init section loses that.

andyross commented:
> What do you find hacky about it?

I think the argument would go that it's still a kludge intended to express "dependencies" between devices that aren't reflected in the existing devicetree data.  And thus (and not trying to put words in @tbursztyka 's mouth, though I think he and I see this similarly) if we're going to pick a kludge, let's pick the simplest one with the least overhead.

Meh.  I don't actually feel strongly here, personally.  But my gut says the extra work being done in the callbacks in this version isn't buying us much.

fabiobaltieri commented:
> Meh. I don't actually feel strongly here, personally. But my gut says the extra work being done in the callbacks in this version isn't buying us much.

It allows to express a "initialize or do not initialize condition" and let the system honor the original priority (which includes the sub-priority/ordinal and is build time validated), which is not possible by the other proposal - and still allows deferred initialization. I think it's a very concrete use case for system with alternate part numbers.

carlescufi commented:
@npitre would you please take a look at this alternative to https://github.com/zephyrproject-rtos/zephyr/pull/67335?

edersondisouza commented:
Closing in favour of #67335
